src/lib/signature.ts

```diff
@@ -208,10 +225,51 @@ class Signature extends CircuitValue {
    */
   verify(publicKey: PublicKey, msg: Field[]): Bool {
     const point = publicKey.toGroup();
-    let e = Scalar.fromBits(
-      Poseidon.hash(msg.concat([point.x, point.y, this.r])).toBits()
+    let h = hashWithPrefix(
+      prefixes.signatureTestnet,
+      msg.concat([point.x, point.y, this.r])
     );
-    let r = point.scale(e).neg().add(Group.generator.scale(this.s));
+    // TODO: Scalar.fromBits interprets the input as a "shifted scalar"
+    // therefore we have to use scaleShifted which is very inefficient
+    let e = Scalar.fromBits(h.toBits());
```

Comment:
is this going to create constraints?

Reply:
yeah sure! `toBits` does the proper in-snark unpacking of a field element, and `fromBits` is a no-op, since scalars are already represented as an array of bits (unfortunately, the bits represent a shifted scalar :/)